File: src/Admin/Migration.Admin.Web/src/api/telemetryEventWriter.ts
import { apiPost } from './core/adminApiClient';
import type { TelemetryWriteResult } from './telemetrySink';

export type TelemetryEventWriteRequest = {
  workspaceId: string;
  eventName: string;
  category: string;
  severity: string;
  tenantId?: string | null;
  projectId?: string | null;
  runId?: string | null;
  correlationId?: string | null;
  dimensions?: Record<string, string> | null;
  metrics?: Record<string, number> | null;
};

export type TelemetryEventWriterProbe = {
  request: TelemetryEventWriteRequest;
  result: TelemetryWriteResult;
};

export async function probeTelemetryEventWriter(): Promise<TelemetryEventWriterProbe> {
  return apiPost<TelemetryEventWriterProbe>('/api/cloud/telemetry/writer/probe', {});
}

File: src/Core/Migration.ControlPlane/Telemetry/TelemetryEventWriter.cs
namespace Migration.ControlPlane.Telemetry;

public sealed class TelemetryEventWriter : ITelemetryEventWriter
{
    private readonly ITelemetrySink _sink;

    public TelemetryEventWriter(ITelemetrySink sink)
    {
        _sink = sink;
    }

    public Task<TelemetryWriteResult> WriteAsync(
        TelemetryEventWriteRequest request,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);

        var telemetryEvent = TelemetryEventFactory.Create(
            workspaceId: request.WorkspaceId,
            eventName: request.EventName,
            category: request.Category,
            severity: request.Severity,
            tenantId: request.TenantId,
            projectId: request.ProjectId,
            runId: request.RunId,
            correlationId: request.CorrelationId,
            dimensions: request.Dimensions,
            metrics: request.Metrics);

        return _sink.WriteAsync(telemetryEvent, cancellationToken);
    }
}
